src/functional/ffront/decorator.py

```diff
@@ -154,8 +155,8 @@ def is_cartesian_offset(o: FieldOffset):
 
 
 def _field_constituents_shape_and_dims(
-    arg, arg_type: ts.FieldType | ts.ScalarType | ts.TupleType
-) -> Generator[tuple[tuple[int, ...], list[Dimension]]]:
+    arg, arg_type: ts.DataType
+) -> Iterator[tuple[Optional[tuple[Any]], list[Dimension]]]:
```

Comment:
this seems to general

Reply:
also, `arg` is missing an annotation.

Reply:
I put ` Any`there, tbh the zip creates issues with types